### ISSUE TITLE
Feature request: LIST/SET data structure

### ISSUE THREAD

**Issue by pcdinh** [Amaury/FineDB#2]
Hi,

At the moment, FineDB supports key-value data model where value is a string or a number.

However, in practice, the above data model makes it hard to use in the wider scope other than simple cache store

There are quite common cases where LIST/SET data structure can be quite handy:
- Most recent read articles/posts in a category or front page or forum: in this case SET is quite helpful to avoid adding duplicate ID into a list of articles or posts. Because the list of IDs is populated by many processes/requests at the same time, server-side serialization should be implemented for efficiency.  
- Activity streams
- Paging for user generated content
- Chat messages
- Room list 
- Room-specific occupants
- Forum inbox/private messages
- Request statistics
  ....

**Comment by pcdinh** [comment]
```
http://www.finedb.org/page/show/1
```

It looks like that LIST/SET requires lock-free data structure to be fast in multiple-threaded FineDB. Also, the protocol must be changed a little bit to support more commands

**Comment by Amaury** [comment]
Hi! Thank you for your interrest in FineDB.
I guess you are thinking about complex data types like those offered by Redis.

Short anwser to your question: No, there is no plan to add support for LIST/SET/HASH in FineDB.

Long answer:
- FineDB uses LMDB as its storage engine. And LMDB is a key/value store.
- Using Redis on professional projects, I started using most of its data types (lists, hashes, sorted sets). But after some time, it seemed too limited (only one depth level, no "sub-hashes" nor lists of lists). I ended writing some JSON-serialized data as values in Redis. It is totally acceptable as long as you don't need introspection on the nested data (or if you define your own indexes).
- As I wrote in my blog (in french, sorry), I'm thinking about storing complex data, serialized using MessagePack. It's the reason of the "serialized data bit" int the FineDB's protocol. Adding this feature could potentially turn FineDB in a document-oriented database. This vision opens interesting perspectives, like map/reduce.
- I have two goals with FineDB:
  - Create the fastest key/value store.
  - Use it to build the simplest scalable platform. (simple to set-up, simple to use)

But I'm open to any argument that could change my opinion.
